### ISSUE TITLE
Support for ARM ldrex/strex Instructions

### ISSUE THREAD

**Issue by pdamian** [JonathanSalwan/Triton#1053]
Would it be possible to add support for (ARM32) instructions such as load-exclusive `ldrex` and store-exclusive `strex`?

**Comment by pdamian** [comment]
I agree to go for solution 2. At least for the usecases I had so far, that should be enough.

**Comment by cnheitman** [comment]
I just pushed a commit for this issue. I implemented a simplified version as discussed.

I had time only to test it with simple and quite specific cases. @pdamian Could you give it a try and let me known if it works correctly on the sample you were analyzing? The implementation can be further improved by adding a flag associated with the load/store addresses but first I would like to have some feedback.

**Comment by JonathanSalwan** [comment]
Nice patch @cnheitman, looks good to me.

> The implementation can be further improved by adding a flag associated with the load/store addresses but first I would like to have some feedback.

Yep it could be nice to have a flag so that we can avoid exporting specific methods for ARM32 (`isMemoryExclusiveAccess` and `setMemoryExclusiveAccess`) into the `Architecture` class. However that is a really good start!

**Comment by pdamian** [comment]
Thanks @cnheitman for implementing. Looks good on my test case!

**Comment by cnheitman** [comment]
Thanks for checking it out, @pdamian.

@JonathanSalwan The question is should we improve this approach or leave this way? Also, regarding your comment on `isMemoryExclusiveAccess`/`setMemoryExclusiveAccess`. I didn't like adding those methods to `Architecture` either. However, adding the mem flag (`exclusive_mem_acc`) into the `Arm32Cpu` class seemed the right choice (and that forced me to update the `Architecture` class). Do you have in mind another approach?

(Btw, I just realized that I switched from camel case to snake case style in the variable naming... I'll fix that in the next commit.)

**Comment by JonathanSalwan** [comment]
@cnheitman I think we can leave this way. Looks good :)

**Comment by JonathanSalwan** [comment]
I think we can close this issue then?

**Comment by cnheitman** [comment]
Yes, we can close it. (The only thing pending on my side is fixing the naming of the variables. I'll push that as soon as I have some spare time.)
